internal/database/migrations/59_add_catalog_item_ref_triggers.up.sql

```
@@ -0,0 +1,187 @@
+-- 1. BEFORE UPDATE triggers on cluster_catalog_items and compute_instance_catalog_items that prevent soft-deleting a
+--    catalog item while active clusters or compute instances still reference it.
+--
+-- 2. BEFORE INSERT OR UPDATE triggers on clusters and compute_instances that prevent creating or updating a resource
+--    that references a catalog item that does not exist or has been soft-deleted. These triggers use FOR SHARE to
+--    acquire a shared lock on the catalog item row, which conflicts with the exclusive lock held by a concurrent
+--    catalog item soft-delete. This bidirectional locking eliminates the TOCTOU race condition.
+
+-- Indexes to speed up the lookup of clusters and compute instances by their catalog_item reference:
+create index clusters_catalog_item on clusters
+  using btree ((data->'spec'->>'catalog_item'))
+  where data->'spec'->>'catalog_item' is not null;
```

SQLFluff (4.2.2)

[error] CREATE INDEX should use CONCURRENTLY to avoid locking the table during the build.

(PG01)